fix(protocol-designer): auto-generate trashBin for flex if no pipetting commands exist

jerader opened the pull request:
closes AUTH-267

# Overview

In PD in production, if you create a Flex protocol with no pipetting steps and reupload it to the app, it does not auto generate a trash bin for you. Technically there is a way around it by adding a trash bin entity. But its a weird state because return tip and no trash entity is not supported in PD. 

Ot-2 does not have this issue because it always ensures a hardcoded fixed trash is generated

So this PR auto-generates a trash bin for a flex if there is none when importing a protocol.

# Test Plan

I created a test case to test all scenarios, so it should work as expected. 

However, you can create a flex protocol and add a pause step. Download the protocol and reimport it back to PD. See that the trash bin is generated.

# Changelog

- create util for getting an unoccupied slot, taking into account modules, labwares, move labware commands, waste chute, and staging area slots
- add test

# Review requests

see test plan

# Risk assessment

low

jerader committed:
fix trash bin auto generating

jerader committed:
fix(protocol-designer): auto-generate trashBin for flex if no pipetting commands exist

closes AUTH-267

jerader committed:
remove unneeded logic for adapters

jerader commented on protocol-designer/src/components/modals/CreateFileWizard/utils.ts:
```diff
@@ -241,13 +241,6 @@ export const getTrashSlot = (values: FormState): string => {
     ? [WASTE_CHUTE_CUTOUT as string]
     : []
 
-  if (
-    !cutouts.includes(FLEX_TRASH_DEFAULT_SLOT) &&
-    !moduleSlots.includes('A3')
-  ) {
-    return FLEX_TRASH_DEFAULT_SLOT
-  }
-
```
now unnecessary logic since we are listing 'A3' as the first cutout in the array to map through

jerader committed:
fix test case text

koji commented on protocol-designer/src/step-forms/utils/index.ts:
```diff
@@ -28,9 +30,55 @@ import type {
+  FormPipette,
+  LabwareOnDeck as LabwareOnDeckType,
+} from '../types'
+import { getCutoutIdByAddressableArea } from '../../utils'
```
this can be up

koji commented on protocol-designer/src/step-forms/utils/index.ts:
```diff
@@ -248,3 +296,84 @@ export function getHydratedForm(
   return hydratedForm
 }
+
+export function getUnoccupiedSlotForMoveableTrash(
```
if there is any specific reason, using an arrow function would be good for consistency.

koji commented on protocol-designer/src/step-forms/utils/index.ts:
```diff
@@ -248,3 +296,84 @@ export function getHydratedForm(
+  stagingAreaSlotNames: AddressableAreaName[]
+): string {
+  const wasteChuteSlot = hasWasteChuteCommands
+    ? [WASTE_CHUTE_CUTOUT as string]
```
```suggestion
    ? [WASTE_CHUTE_CUTOUT]
```

koji commented on protocol-designer/src/step-forms/utils/index.ts:
```diff
@@ -248,3 +296,84 @@ export function getHydratedForm(
+    return ''
+  }
+
+  return unoccupiedSlot?.slot
```
this line is after null-check so `?` is not required.

```suggestion
  return unoccupiedSlot.slot
```

koji commented:
question

when I added waste chute, export and import the protocol, trash bin is generated but waste chute is gone.
Is this expected behavior?

jerader commented:
> > question
> > when I added waste chute, export and import the protocol, trash bin is generated but waste chute is gone. Is this expected behavior?
> 
> @koji oops that is definitely a bug! Thanks for catching it.

@koji actually, if you did not add any waste chute commands, it is not a bug. But if you did add waste chute commands, it is a bug.

koji commented:
> > > question
> > > when I added waste chute, export and import the protocol, trash bin is generated but waste chute is gone. Is this expected behavior?
> > 
> > 
> > @koji oops that is definitely a bug! Thanks for catching it.
> 
> @koji actually, if you did not add any waste chute commands, it is not a bug since we remove any staging area slot and waste chute that is not used in the protocol. But if you did add waste chute commands, it is a bug.

@jerader I added waste chute from the initial setup wizard.

koji commented:
one more question

I compared two json.
left: exported json - no trash bin
right: import the left json and re-export (trash bin with auto generated)

making pipetteTiprackAssignments item nested (adding `[]`) is an expected behavior?

<img width="1271" alt="Screenshot 2024-04-10 at 1 57 51 PM" src="https://github.com/Opentrons/opentrons/assets/474225/2012d972-f364-427d-afab-1f58bc9417de">

left
[no-trashbin.json](https://github.com/Opentrons/opentrons/files/14935992/no-trashbin.json)

right
[no-trashbin (1).json](https://github.com/Opentrons/opentrons/files/14935993/no-trashbin.1.json)

jerader commented:
> one more question
> 
> I compared two json. left: exported json - no trash bin right: import the left json and re-export (trash bin with auto generated)
> 
> making pipetteTiprackAssignments item nested (adding `[]`) is an expected behavior?
> 
> left [no-trashbin.json](https://github.com/Opentrons/opentrons/files/14935992/no-trashbin.json)
> 
> right [no-trashbin (1).json](https://github.com/Opentrons/opentrons/files/14935993/no-trashbin.1.json)

@koji  oh i think what's happening is when you reimport, it goes through the 8.1 migration because the version when you export is `8.0.3` and not `8.1.0` - this is a current limitation with PD where the versioning doesn't update until i push a release tag 😦 so for now we have to manually change the version before we reimport.

jerader commented:
> > > > question
> > > > when I added waste chute, export and import the protocol, trash bin is generated but waste chute is gone. Is this expected behavior?
> > > 
> > > 
> > > @koji oops that is definitely a bug! Thanks for catching it.
> > 
> > 
> > @koji actually, if you did not add any waste chute commands, it is not a bug since we remove any staging area slot and waste chute that is not used in the protocol. But if you did add waste chute commands, it is a bug.
> 
> @jerader I added waste chute from the initial setup wizard.

@koji got it, this sounds like expected behavior. if you didn't add any step using the waste chute then it isn't being used in the protocol and won't generate upon reimport or even if you upload it to the app.

jerader committed:
address feedback

koji reviewed (approved):
thank you for answering to all my questions.
LGTM!
